### PR TITLE
Update FileSystem UseWinRT heuristic

### PR DESCRIPTION
**Pull request by ericstj**
This improves the performance and correctness of the UseWinRT heuristic.  Details are in individual commits.  These problems were found in testing of IsolatedStorage and actual store apps.

### CODE REVIEW

File: src/System.IO.FileSystem/src/System/IO/MultiplexingWin32WinRTFileSystem.cs [hunk 1]
```diff
@@ -93,7 +94,8 @@ public override void MoveFile(string sourceFullPath, string destFullPath)
 
         public override FileStreamBase Open(string fullPath, FileMode mode, FileAccess access, FileShare share, int bufferSize, FileOptions options, FileStream parent)
         {
-            return Select(fullPath, isCreate: true).Open(fullPath, mode, access, share, bufferSize, options, parent);
+            bool isCreate = mode != FileMode.Open && mode != FileMode.Truncate;
```
Comment (stephentoub) [inline src/System.IO.FileSystem/src/System/IO/MultiplexingWin32WinRTFileSystem.cs]: Does FileMode.OpenOrCreate matter here?

Comment (ericstj) [inline src/System.IO.FileSystem/src/System/IO/MultiplexingWin32WinRTFileSystem.cs]: Open or create will fall back to create if the file doesn't exist, so we need to treat it like a create.

File: src/System.IO.FileSystem/src/System/IO/MultiplexingWin32WinRTFileSystem.cs [hunk 2]
```diff
@@ -145,33 +147,63 @@ private bool ShouldUseWinRT(string fullPath, bool isCreate)
 
             do
             {
-                Interop.WIN32_FIND_DATA findData = new Interop.WIN32_FIND_DATA();
-                using (SafeFindHandle handle = Interop.mincore.FindFirstFile(fullPath, ref findData))
-                {
-                    int error = Marshal.GetLastWin32Error();
+                int error = Interop.ERROR_SUCCESS;
```
Comment (stephentoub) [inline src/System.IO.FileSystem/src/System/IO/MultiplexingWin32WinRTFileSystem.cs]: Can you add a comment to ShouldUseWinRT indicating what determines whether we should use WinRT or not? The method is sufficiently complicated that it's not clear on first-glance what the method is really looking for to make its decision.

File: src/System.IO.FileSystem/src/System/IO/MultiplexingWin32WinRTFileSystem.cs [hunk 3]
```diff
@@ -145,33 +147,63 @@ private bool ShouldUseWinRT(string fullPath, bool isCreate)
...
+                            {
+                                // got the find data, use WinRT for placeholder files
+
+                                Debug.Assert((findData.dwFileAttributes & Interop.FILE_ATTRIBUTE_DIRECTORY) == 0);
+                                Debug.Assert((findData.dwFileAttributes & Interop.FILE_ATTRIBUTE_REPARSE_POINT) != 0);
+
+                                useWinRt = findData.dwReserved0 == Interop.IO_REPARSE_TAG_FILE_PLACEHOLDER;
+                                break;
+                            }
+                            else
```
Comment (stephentoub) [inline src/System.IO.FileSystem/src/System/IO/MultiplexingWin32WinRTFileSystem.cs]: We're breaking out of the loop on all success cases.  As such, can you just remove the else clauses and put ```int error = Marshal.GetLastWin32Error()``` down where you have the ```Debug.Assert(error != Interop.ERROR_SUCCESS```?

**Comment by stephentoub** [comment]
Other than my minor comments, LGTM.

File: src/System.IO.FileSystem/src/System/IO/MultiplexingWin32WinRTFileSystem.cs (hunk 3 above)
Comment (ericstj) [inline src/System.IO.FileSystem/src/System/IO/MultiplexingWin32WinRTFileSystem.cs]: Sure, I considered this initially but I'm sensitive to folks making inadvertent changes that blow away GLE.  I suppose the Assert should help protect us here.

File: src/System.IO.FileSystem/src/System/IO/MultiplexingWin32WinRTFileSystem.cs (hunk 2 above)
Comment (ericstj) [inline src/System.IO.FileSystem/src/System/IO/MultiplexingWin32WinRTFileSystem.cs]: Sure, will add a comment.

**Comment by ericstj** [comment]
Thanks, I addressed those comments.  Will commit after build succeeds.

**Comment by ericstj** [comment]
Talked to PallaviT, process test issue is known and being fixed.  @dotnet-bot test this please

**Comment by ericstj** [comment]
Same process test is failing @pallavit.  Proceeding with merge.

**Pull request merged**